### PR TITLE
Feature/1016

### PR DESCRIPTION
**Pull request by ThanhD-AVISPL**
This pull request implements: 

- Hub monitoring and its unit test.
- Aggregator (device dashboard, edit location, edit room, create room, scenes trigger), and its unit test.
- Aggregated device and its unit test

### CODE REVIEW

File: src/main/java/com/avispl/symphony/dal/infrastructure/management/samsung/smartthings/SamsungSmartThingsAggregatorCommunicator.java [hunk 1]
```diff
@@ -0,0 +1,3122 @@
...
+			DeviceWrapper responseDeviceList = this.doGet(request, DeviceWrapper.class);
+
+			if (responseDeviceList.getDevices() != null) {
+				cachedDevices = (ConcurrentHashMap<String, Device>) responseDeviceList.getDevices().stream().collect(Collectors.toConcurrentMap(Device::getDeviceId, Function.identity()));
+			}
+		} catch (Exception e) {
+			if (retryOnError) {
+				retrieveDevices(false);
+			}
+			logger.error(String.format("Aggregated Device Data Retrieval-Error: %s with cause: %s", e.getMessage(), e.getCause().getMessage()), e);
```
Comment (MaksimR-AVISPL) [inline src/main/java/com/avispl/symphony/dal/infrastructure/management/samsung/smartthings/SamsungSmartThingsAggregatorCommunicator.java:1317]: maybe would be nice to sneak in locationId here, to see what request this was exactly right off the bat

File: src/main/java/com/avispl/symphony/dal/infrastructure/management/samsung/smartthings/SamsungSmartThingsAggregatorCommunicator.java [hunk 2]
```diff
@@ -0,0 +1,3122 @@
...
+				cachedLocations.set(locationIndex, location);
+			} else {
+				throw new ResourceNotReachableException(String.format("Changing %s name fail, please try again later", controllableProperty));
+			}
+
+			addAdvanceControlProperties(advancedControllableProperties,
+					createText(stats, AggregatorGroupControllingMetric.LOCATION_MANAGEMENT.getName() + LocationManagementMetric.LOCATION.getName() + locationIndex, value));
+			isEmergencyDelivery = true;
+		} catch (Exception e) {
+			throw new ResourceNotReachableException(String.format("Error while controlling location %s: %s", controllableProperty, e.getMessage()), e);
```
Comment (MaksimR-AVISPL) [inline src/main/java/com/avispl/symphony/dal/infrastructure/management/samsung/smartthings/SamsungSmartThingsAggregatorCommunicator.java:1374]: This error could also use some specifics like locationName for instance. Would make it a bit more clear at glance, plus we can't be sure that error message would include that

File: src/main/java/com/avispl/symphony/dal/infrastructure/management/samsung/smartthings/SamsungSmartThingsAggregatorCommunicator.java [hunk 3]
```diff
@@ -0,0 +1,3122 @@
...
+	}
+
+	/**
+	 * Handle rate limit exceed error while controlling
+	 *
+	 * @param response ResponseEntity
+	 * @throws ResourceNotReachableException when get limit rate exceed error
+	 */
+	private void handleRateLimitExceed(ResponseEntity<?> response) {
+		Optional<String> rateLimit = Optional.ofNullable(response)
```
Comment (MaksimR-AVISPL) [inline src/main/java/com/avispl/symphony/dal/infrastructure/management/samsung/smartthings/SamsungSmartThingsAggregatorCommunicator.java:2901]: Is it possible to achieve this with the http interceptor? Similarly to the way 429 is handled in XiO aggregator

**Review by MaksimR-AVISPL — CHANGES_REQUESTED** [review]

File: pom.xml [hunk 4]
```diff
@@ -0,0 +1,223 @@
...
+            <groupId>com.google.guava</groupId>
+            <artifactId>guava</artifactId>
+            <version>30.1.1-jre</version>
+            <scope>provided</scope>
+        </dependency>
+
+        <!--test dependencies-->
+        <dependency>
+            <groupId>org.snmp4j</groupId>
+            <artifactId>snmp4j</artifactId>
```
[the remark this answers is not in the document]
Comment (ThanhD-AVISPL) [reply pom.xml:182]: I removed the redundant snmp4j

File: src/main/java/com/avispl/symphony/dal/infrastructure/management/samsung/smartthings/common/AggregatedDeviceControllingMetric.java [hunk 5]
```diff
@@ -0,0 +1,59 @@
...
+	 * Parameterized constructor
+	 *
+	 * @param name Name of Hub monitoring metric
+	 */
+	AggregatedDeviceControllingMetric(String name) {
+		this.name = name;
+	}
+
+	/**
+	 * retrieve {@code {@link #name}}
```
[the remark this answers is not in the document]
Comment (ThanhD-AVISPL) [reply src/main/java/com/avispl/symphony/dal/infrastructure/management/samsung/smartthings/common/AggregatedDeviceControllingMetric.java:41]: I fixed it

File: src/test/java/com/avispl/symphony/dal/infrastructure/management/samsung/smartthings/SamsungSmartThingsAggregatorCommunicatorTest.java [hunk 6]
```diff
@@ -0,0 +1,844 @@
...
+
+		Assertions.assertEquals(propertyValue, stats.get(propertyName));
+	}
+
+	/**
+	 * Test SamSungSmartThingsAggregator.controlProperty aggregated device: saturation control
+	 *
+	 * Expected: control successfully
+	 */
+	@Tag("RealDevice")
```
[the remark this answers is not in the document]
Comment (ThanhD-AVISPL) [reply src/test/java/com/avispl/symphony/dal/infrastructure/management/samsung/smartthings/SamsungSmartThingsAggregatorCommunicatorTest.java:809]: I fixed it

File: src/main/java/com/avispl/symphony/dal/infrastructure/management/samsung/smartthings/dto/room/Room.java [hunk 7]
```diff
@@ -0,0 +1,105 @@
...
+ */
+package com.avispl.symphony.dal.infrastructure.management.samsung.smartthings.dto.room;
+
+import com.fasterxml.jackson.annotation.JsonAlias;
+import com.fasterxml.jackson.annotation.JsonIgnoreProperties;
+
+import com.avispl.symphony.dal.util.StringUtils;
+
+/**
+ * Rom
```
[the remark this answers is not in the document]
Comment (ThanhD-AVISPL) [reply src/main/java/com/avispl/symphony/dal/infrastructure/management/samsung/smartthings/dto/room/Room.java:12]: yes. I fixed it

File: src/main/java/com/avispl/symphony/dal/infrastructure/management/samsung/smartthings/SamsungSmartThingsAggregatorCommunicator.java (hunk 2 above)
Comment (ThanhD-AVISPL) [reply src/main/java/com/avispl/symphony/dal/infrastructure/management/samsung/smartthings/SamsungSmartThingsAggregatorCommunicator.java:1374]: I fixed it

File: src/main/java/com/avispl/symphony/dal/infrastructure/management/samsung/smartthings/SamsungSmartThingsAggregatorCommunicator.java (hunk 1 above)
Comment (ThanhD-AVISPL) [reply src/main/java/com/avispl/symphony/dal/infrastructure/management/samsung/smartthings/SamsungSmartThingsAggregatorCommunicator.java:1317]: I fixed it

File: src/main/java/com/avispl/symphony/dal/infrastructure/management/samsung/smartthings/SamsungSmartThingsAggregatorCommunicator.java (hunk 3 above)
Comment (ThanhD-AVISPL) [reply src/main/java/com/avispl/symphony/dal/infrastructure/management/samsung/smartthings/SamsungSmartThingsAggregatorCommunicator.java:2901]: The adapter will reach the rate limit when we send too many control requests in a minute. I think we don't need to build the HTTP interceptor.

File: src/main/java/com/avispl/symphony/dal/infrastructure/management/samsung/smartthings/common/AggregatedDeviceControllingMetric.java [hunk 8]
```diff
@@ -0,0 +1,58 @@
+/*
+ * Copyright (c) 2022 AVI-SPL, Inc. All Rights Reserved.
+ */
+package com.avispl.symphony.dal.infrastructure.management.samsung.smartthings.common;
+import java.util.Arrays;
+import java.util.Optional;
+
+/**
+ * Aggregated Device Controlling Metric
```
Comment (MaksimR-AVISPL) [inline src/main/java/com/avispl/symphony/dal/infrastructure/management/samsung/smartthings/common/AggregatedDeviceControllingMetric.java:9]: Ok this isn't what I meant exactly :) This description is sort of done by the class name already, what I mean is... Javadoc description that would explain what purpose does it serve exactly for someone who's out of context (like Bohdan, who was asked to review adapters code before, and who has never seen what adapters really are).

File: src/main/java/com/avispl/symphony/dal/infrastructure/management/samsung/smartthings/dto/presentation/DetailViewPresentation.java [hunk 9]
```diff
@@ -0,0 +1,234 @@
...
+import com.fasterxml.jackson.annotation.JsonIgnoreProperties;
+
+import com.avispl.symphony.dal.infrastructure.management.samsung.smartthings.dto.presentation.controllableproperties.type.DropdownList;
+import com.avispl.symphony.dal.infrastructure.management.samsung.smartthings.dto.presentation.controllableproperties.type.PushButton;
+import com.avispl.symphony.dal.infrastructure.management.samsung.smartthings.dto.presentation.controllableproperties.type.Slider;
+import com.avispl.symphony.dal.infrastructure.management.samsung.smartthings.dto.presentation.controllableproperties.type.Switch;
+import com.avispl.symphony.dal.infrastructure.management.samsung.smartthings.dto.presentation.controllableproperties.type.TextOrNumber;
+
+/**
+ * Detail View Presentation define how the SmartThings app (or other clients) should present the Attributes and Commands of a Device in detail view/ Aggregated Device view
```
Comment (MaksimR-AVISPL) [inline src/main/java/com/avispl/symphony/dal/infrastructure/management/samsung/smartthings/dto/presentation/DetailViewPresentation.java:16]: For other DTO/constants/enum classes - similar description, that would really give a high-level understanding of its purpose would work great :)

File: src/main/java/com/avispl/symphony/dal/infrastructure/management/samsung/smartthings/SamsungSmartThingsAggregatorCommunicator.java [hunk 10]
```diff
@@ -0,0 +1,3177 @@
...
+									.map(u -> u.get(SmartThingsConstant.VALUE))
+									.map(JsonNode::asText)
+									.orElse(SmartThingsConstant.NONE);
+							devicePresentation.getColor().setSaturation(Float.parseFloat(value));
+							continue;
+						}
+					}
+					break;
+				default:
+					break;
```
Comment (MaksimR-AVISPL) [inline src/main/java/com/avispl/symphony/dal/infrastructure/management/samsung/smartthings/SamsungSmartThingsAggregatorCommunicator.java:340]: the question above + a log entry here, indicating that there's no expected behavior for this exist

File: src/main/java/com/avispl/symphony/dal/infrastructure/management/samsung/smartthings/SamsungSmartThingsAggregatorCommunicator.java [hunk 11]
```diff
@@ -0,0 +1,3177 @@
...
+		String locationName = cachedLocations.stream().filter(l -> l.getLocationId().equals(locationIdFiltered)).map(Location::getName).findFirst()
+				.orElse(cachedLocations.get(0).getName());
+		stats.put(HubInfoMetric.CURRENT_LOCATION.getName(), locationName);
+	}
+
+	/**
+	 * This method is used to retrieve rooms info by send GET request to "https://api.smartthings.com/v1/locations/{locationId}/rooms"
+	 *
+	 * @param retryOnError retry if any error occurs
+	 * @throws ResourceNotReachableException When getting the empty response
```
Comment (MaksimR-AVISPL) [inline src/main/java/com/avispl/symphony/dal/infrastructure/management/samsung/smartthings/SamsungSmartThingsAggregatorCommunicator.java:1120]: it doesn't seem like this method really throws anything

File: src/main/java/com/avispl/symphony/dal/infrastructure/management/samsung/smartthings/SamsungSmartThingsAggregatorCommunicator.java [hunk 12]
```diff
@@ -0,0 +1,3177 @@
...
+			stats.put(AggregatorGroupControllingMetric.DEVICES_DASHBOARD.getName() + AggregatorGroupControllingMetric.MESSAGE.getName(), "There are no devices available in this room or location");
+		} else {
+			stats.remove(AggregatorGroupControllingMetric.DEVICES_DASHBOARD.getName() + AggregatorGroupControllingMetric.MESSAGE.getName());
+		}
+	}
+
+	/**
+	 * This method is used to validate if the locations contain the No room Assigned
+	 *
+	 */
```
Comment (MaksimR-AVISPL) [inline src/main/java/com/avispl/symphony/dal/infrastructure/management/samsung/smartthings/SamsungSmartThingsAggregatorCommunicator.java:1662]: missing @return tag. Also, is there a reason this method is public?

**Review by MaksimR-AVISPL — CHANGES_REQUESTED** [review]

File: src/main/java/com/avispl/symphony/dal/infrastructure/management/samsung/smartthings/SamsungSmartThingsAggregatorCommunicator.java (hunk 11 above)
Comment (ThanhD-AVISPL) [reply src/main/java/com/avispl/symphony/dal/infrastructure/management/samsung/smartthings/SamsungSmartThingsAggregatorCommunicator.java:1120]: I fixed it

File: src/main/java/com/avispl/symphony/dal/infrastructure/management/samsung/smartthings/SamsungSmartThingsAggregatorCommunicator.java (hunk 12 above)
Comment (ThanhD-AVISPL) [reply src/main/java/com/avispl/symphony/dal/infrastructure/management/samsung/smartthings/SamsungSmartThingsAggregatorCommunicator.java:1662]: I fixed it

File: src/main/java/com/avispl/symphony/dal/infrastructure/management/samsung/smartthings/SamsungSmartThingsAggregatorCommunicator.java (hunk 10 above)
Comment (ThanhD-AVISPL) [reply src/main/java/com/avispl/symphony/dal/infrastructure/management/samsung/smartthings/SamsungSmartThingsAggregatorCommunicator.java:340]: I used to map for some unique controllable properties. if it's common control, we'll map it in the next step. I'm going to take a note here "we do not log error/ warning here, we are going to map data for common controllable properties in the next step"

File: src/main/java/com/avispl/symphony/dal/infrastructure/management/samsung/smartthings/dto/room/RoomWrapper.java [hunk 13]
```diff
@@ -0,0 +1,42 @@
...
+package com.avispl.symphony.dal.infrastructure.management.samsung.smartthings.dto.room;
+
+import java.util.ArrayList;
+import java.util.List;
+
+import com.fasterxml.jackson.annotation.JsonAlias;
+import com.fasterxml.jackson.annotation.JsonIgnoreProperties;
+
+/**
+ * RoomWrapper
```
[the remark this answers is not in the document]
Comment (ThanhD-AVISPL) [reply src/main/java/com/avispl/symphony/dal/infrastructure/management/samsung/smartthings/dto/room/RoomWrapper.java]: I fixed it

File: src/main/java/com/avispl/symphony/dal/infrastructure/management/samsung/smartthings/dto/presentation/DetailViewPresentation.java (hunk 9 above)
Comment (ThanhD-AVISPL) [reply src/main/java/com/avispl/symphony/dal/infrastructure/management/samsung/smartthings/dto/presentation/DetailViewPresentation.java:16]: I fixed the other DTO like this one

File: src/main/java/com/avispl/symphony/dal/infrastructure/management/samsung/smartthings/common/AggregatedDeviceControllingMetric.java (hunk 8 above)
Comment (ThanhD-AVISPL) [reply src/main/java/com/avispl/symphony/dal/infrastructure/management/samsung/smartthings/common/AggregatedDeviceControllingMetric.java:9]: I fixed it

File: src/main/java/com/avispl/symphony/dal/infrastructure/management/samsung/smartthings/SamsungSmartThingsAggregatorCommunicator.java [hunk 14]
```diff
@@ -0,0 +1,3122 @@
...
+		// Finished collecting
+	}
+
+	//region retrieve detail aggregated device info: device health, device presentation and device full status in worker thread
+	//--------------------------------------------------------------------------------------------------------------------------------
+
+	/**
+	 * @param currentThread current thread index
+	 *
+	 * Submit thread to get device detail info
```
[the remark this answers is not in the document]
Comment (ThanhD-AVISPL) [reply src/main/java/com/avispl/symphony/dal/infrastructure/management/samsung/smartthings/SamsungSmartThingsAggregatorCommunicator.java:140]: I fixed it

File: src/main/java/com/avispl/symphony/dal/infrastructure/management/samsung/smartthings/SamsungSmartThingsAggregatorCommunicator.java [hunk 15]
```diff
@@ -0,0 +1,3122 @@
...
+			logger.error(String.format("Error while retrieve %s health info: %s ", cachedDevicesAfterPollingInterval.get(deviceID).getName(), e.getMessage()), e);
+		}
+	}
+
+	/**
+	 * This method is used to retrieve list of device health by send GET request to
+	 * https://api.smartthings.com/v1/presentation?presentationId={presentationId}&manufacturerName={manufacturerName}
+	 *
+	 * @param deviceId device ID
+	 * @throws ResourceNotReachableException If any error occurs
```
[the remark this answers is not in the document]
Comment (ThanhD-AVISPL) [reply src/main/java/com/avispl/symphony/dal/infrastructure/management/samsung/smartthings/SamsungSmartThingsAggregatorCommunicator.java]: I fixed it

File: src/main/java/com/avispl/symphony/dal/infrastructure/management/samsung/smartthings/SamsungSmartThingsAggregatorCommunicator.java [hunk 16]
```diff
@@ -0,0 +1,3122 @@
...
+			logger.error(String.format("Error while retrieve %s presentation info: %s", device.getName(), e.getMessage()), e);
+		}
+	}
+
+	/**
+	 * This method is used to retrieve device full status by send GET request to
+	 * https://api.smartthings.com/v1/devices/{deviceId}/status
+	 *
+	 * @param deviceId device ID
+	 * @throws ResourceNotReachableException If any error occurs
```
[the remark this answers is not in the document]
Comment (ThanhD-AVISPL) [reply src/main/java/com/avispl/symphony/dal/infrastructure/management/samsung/smartthings/SamsungSmartThingsAggregatorCommunicator.java]: I fixed it

File: src/main/java/com/avispl/symphony/dal/infrastructure/management/samsung/smartthings/SamsungSmartThingsAggregatorCommunicator.java [hunk 17]
```diff
@@ -0,0 +1,3122 @@
...
+					.map(Iterator::next)
+					.map(c -> c.get(detailViewPresentation.getCapability()))
+					.map(JsonNode::elements);
+
+			String value = SmartThingsConstant.NONE;
+			String unit = SmartThingsConstant.NONE;
+
+			// map specific controllable property status to device
+			switch (detailViewPresentation.getCapability()) {
+				case AggregatedDeviceColorControllingConstant.COLOR_CONTROL:
```
[the remark this answers is not in the document]
Comment (ThanhD-AVISPL) [reply src/main/java/com/avispl/symphony/dal/infrastructure/management/samsung/smartthings/SamsungSmartThingsAggregatorCommunicator.java:312]: yes. it is better than if we use if else, but I used switch-case for the extending.

**Review by MaksimR-AVISPL — APPROVED** [review]
Approved, please make sure to reapply getters/setters templates and regenerate getters/setters javadocs :) 

/**
* Retrieves {@link #$field.name}
*
* @return value of {@link #$field.name}
*/
#if($field.modifierStatic)
static ##
#end
$field.type ##
#if($field.recordComponent)
  ${field.name}##
#else
#set($name = $StringUtil.capitalizeWithJavaBeanConvention($StringUtil.sanitizeJavaIdentifier($helper.getPropertyName($field, $project))))
#if ($field.boolean && $field.primitive)
  is##
#else
  get##
#end
${name}##
#end
() {
  return $field.name;
}
/**
* Sets {@link #$field.name} value
*
* @param $field.name new value of {@link #$field.name}
*/
#set($paramName = $helper.getParamName($field, $project))
#if($field.modifierStatic)
static ##
#end
void set$StringUtil.capitalizeWithJavaBeanConvention($StringUtil.sanitizeJavaIdentifier($helper.getPropertyName($field, $project)))($field.type $paramName) {
#if ($field.name == $paramName)
    #if (!$field.modifierStatic)
    this.##
    #else
        $classname.##
    #end
#end
$field.name = $paramName;
}

**Pull request merged**